File: Backup/JUtil.Web/Extensions/ExtGridviewRow.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Collections.Specialized;

namespace JUtil.Web.Extensions
{
    /// <summary>Enhance GridviewRow functionality</summary>
    public static class ExtGridviewRow
    {

        /// <summary>
        /// get button from GridViewRow relay on Button's CommandName
        /// </summary>
        /// <param name="aRow">GridviewRow</param>
        /// <param name="sCommandName">name of Button (maybe it's same as id , not sure !!)</param>
        /// <returns>return null if aRow is not in DataRow mode</returns>
        public static Button GetButton(this GridViewRow aRow, string sCommandName)
        {
            if (aRow.RowType != DataControlRowType.DataRow)
            {
                return null;
            }

            Control oControl = aRow.Cells[0];
            Control oChildControl = null;
            foreach (Control oChildControl_loopVariable in oControl.Controls)
            {
                oChildControl = oChildControl_loopVariable;
                if ((oChildControl) is Button)
                {
                    if (string.Compare(((Button)oChildControl).CommandName, sCommandName, true) == 0)
                    {
                        return oChildControl as Button;
                    }
                }
            }
            return null;
        }

        /// <summary>
        /// set button's text from GridViewRow relay on Button's CommandName
        /// </summary>
        public static void SetButtonText(this GridViewRow aRow, string sCommandName, string txt)
        {
            Button btn = GetButton(aRow, sCommandName);
            if (btn != null)
            {
                btn.Text = txt;
            }
        }

        /// <summary>Extract GridViewRow's Values to an OrderedDictionary</summary>
        public static OrderedDictionary ExtractValues(this GridViewRow aRow)
        {
            if (aRow.RowType != DataControlRowType.DataRow)
            {
                return null;
            }

            GridView grv = aRow.Parent.Parent as GridView;
            DataControlFieldCollection columns = grv.Columns;
            int nColumns = columns.Count;
            OrderedDictionary Result = new OrderedDictionary(nColumns);

            for (int i = 0; i <= nColumns - 1; i++)
            {
                DataControlField aColumn = columns[i];

                if ((aColumn is CommandField))
                    continue;

                OrderedDictionary oDictionary = new OrderedDictionary();
                if ((aColumn is TemplateField))
                {
                    ((TemplateField)aColumn).ExtractValuesFromCell(oDictionary, aRow.Cells[i] as DataControlFieldCell, aRow.RowState, true);
                }
                else
                {
                    aColumn.ExtractValuesFromCell(oDictionary, aRow.Cells[i] as DataControlFieldCell, aRow.RowState, true);
                }

                foreach (DictionaryEntry entry in oDictionary)
                {
                    Result[entry.Key] = entry.Value;
                }
            }
            return Result;
        }
    }
}

File: Backup/Extension/ExtGridviewRow.vb
Imports System.Runtime.CompilerServices
Imports System.Web.UI
Imports System.Web.UI.WebControls
Imports System.Collections.Specialized

''' <summary>Enhance GridviewRow functionality</summary>
Public Module ExtGridviewRow

    ''' <summary>
    ''' get button from GridViewRow relay on Button's CommandName
    ''' </summary>
    ''' <param name="aRow">GridviewRow</param>
    ''' <param name="sCommandName">name of Button (maybe it's same as id , not sure !!)</param>
    ''' <returns>return null if aRow is not in DataRow mode</returns>
    <Extension()> _
    Public Function GetButton(ByVal aRow As GridViewRow, ByVal sCommandName As String) As Button
        If aRow.RowType <> DataControlRowType.DataRow Then
            Return Nothing
        End If

        Dim oControl As Control = aRow.Cells(0)
        Dim oChildControl As Control
        For Each oChildControl In oControl.Controls
            If TypeOf (oChildControl) Is Button Then
                If String.Compare(CType(oChildControl, Button).CommandName, sCommandName, True) = 0 Then
                    Return oChildControl
                End If
            End If
        Next
        Return Nothing
    End Function

    ''' <summary>
    ''' set button's text from GridViewRow relay on Button's CommandName
    ''' </summary>
    <Extension()> _
    Public Sub SetButtonText(ByVal aRow As GridViewRow, ByVal sCommandName As String, ByVal txt As String)
        Dim btn As Button = GetButton(aRow, sCommandName)
        If btn IsNot Nothing Then
            btn.Text = txt
        End If
    End Sub

    ''' <summary>Extract GridViewRow's Values to an OrderedDictionary</summary>
    <Extension()> _
    Public Function ExtractValues(ByVal aRow As GridViewRow) As OrderedDictionary
        If aRow.RowType <> DataControlRowType.DataRow Then
            Return Nothing
        End If

        Dim grv As GridView = aRow.Parent.Parent
        Dim columns As DataControlFieldCollection = grv.Columns
        Dim nColumns As Integer = columns.Count
        Dim Result As OrderedDictionary = New OrderedDictionary(nColumns)

        For i As Integer = 0 To nColumns - 1
            Dim aColumn As DataControlField = columns.Item(i)

            If (TypeOf aColumn Is CommandField) Then Continue For

            Dim oDictionary As OrderedDictionary = New OrderedDictionary()
            If (TypeOf aColumn Is TemplateField) Then
                CType(aColumn, TemplateField).ExtractValuesFromCell(oDictionary, TryCast(aRow.Cells.Item(i), DataControlFieldCell), aRow.RowState, True)
            Else
                aColumn.ExtractValuesFromCell(oDictionary, TryCast(aRow.Cells.Item(i), DataControlFieldCell), aRow.RowState, True)
            End If

            For Each entry As DictionaryEntry In oDictionary
                Result.Item(entry.Key) = entry.Value
            Next
        Next
        Return Result
    End Function
End Module
